pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java

```diff
@@ -461,6 +462,34 @@ public synchronized PinotResourceManagerResponse updateInstanceTags(String insta
     return PinotResourceManagerResponse.SUCCESS;
   }
 
+  /**
+   * Validates whether an instance is offline.
+   * Since ZNodes under "/LIVEINSTANCES" are ephemeral, if there is a ZK session expire (e.g. due to network issue),
+   * the ZNode under "/LIVEINSTANCES" will be deleted. Thus, such race condition can happen when this task is running.
+   * In order to double confirm the live status of an instance, the field "LAST_OFFLINE_TIME" in ZNode under
+   * "/INSTANCES/<instance_id>/HISTORY" needs to be checked. If the value is "-1", that means the instance is ONLINE;
+   * if the value is a timestamp, that means the instance starts to be OFFLINE since that time.
+   * @param instanceId instance id
+   * @param offlineTimeRangeMs the time range that it's valid for an instance to be offline
+   */
+  public boolean isInstanceOffline(String instanceId, long offlineTimeRangeMs) {
```

Comment:
```suggestion
  public boolean shouldInstanceBeRemoved(String instanceId, long offlineTimeRangeMs) {
```

Reply:
I think it's okay to keep the method name as it is, since it's actually checking whether the instance is truly offline. And this method can be reused for other purposes in the future. Dropping instance is only one of the use cases.

Reply:
No, it does more than just check for offline. it checks for threshold of offline time. Maybe change it to `isOfflineForTooLong` or something like that, but I will leave it to you.